db/python/layers/ourdna/dashboard.py

```diff
@@ -0,0 +1,269 @@
+        projects: set[int]
+        samples_internal: list[SampleInternal] = []
+        samples: list[Sample] = []
+        participants: list[ParticipantInternal] = []
+        sample_participants: list[ParticipantInternal] = []
+
+        # Data to be returned
+        collection_to_process_end_time: dict[str, int] = {}
+        collection_to_process_end_time_statistics: dict[str, float | None] = {}
+        collection_to_process_end_time_24h: dict[str, int] = {}
+        processing_times_by_site: dict[str, dict[int, int]] = defaultdict(
+            lambda: defaultdict(int)
+        )  # {site: {hour_bucket: count}}
+        total_samples_by_collection_event_name: dict[str, int] = defaultdict(int)
+        samples_lost_after_collection: dict[str, dict[str, Any]] = {}
+        samples_concentration_gt_1ug: dict[str, float] = {}
+        participants_consented_not_collected: list[int] = []
+        participants_signed_not_consented: list[int] = []
+
+        # TODO We should figure out if we need to handle more than a single project_id
+        # for project_id in project_ids:
+        participants.extend(
+            await self.participant_table.get_participants(project=project_id)
+        )
+
+        projects, samples_internal = await self.sample_table.query(filter_=filter_)
+        samples = [s.to_external() for s in samples_internal]
+
+        if check_project_ids:
+            await self.ptable.check_access_to_project_ids(
+                self.author, projects, readonly=True
+            )
+
+        for participant in participants:
+            samples_for_participant = [
+                sample for sample in samples if sample.participant_id == participant.id
+            ]
+
+            # Create a new participant object with the samples for this participant
+            sample_participants.append(
+                participant.copy(update={'samples': samples_for_participant})
+            )
+
+            # Get the samples that have been collected and consented
+            samples_consented_not_collected = [
+                sample
+                for sample in samples_for_participant
+                if participant.meta.get('consent')
+                and self.get_meta_property(sample=sample, property_name='collection-time') is None
+            ]
+
+            if len(samples_consented_not_collected) > 0:
+                participants_consented_not_collected.append(participant.id)
+
+            # Get the participants that have signed but not consented
+            if participant.meta.get('consent') is None:
+                participants_signed_not_consented.append(participant.id)
+
+        # TODO Add logic here to query and aggregate the data
+
+        # go through all samples and get the meta and do something with it
+        # if we want stats for samples by participant, move this logic up before participant and make use of
+        # the filtered samples to map to each participant
+        for sample in samples:
```

Comment:
Could you move this to a separate function, and return some data structure that contains all these stats? There's a lot of logic in here that IMO could also be broken into separate function.

A small note, there's very little computational difference between these two, so consider that when this gets refactored:

```python
for sample in samples:
    if sample.condition_1:
        list1.append(sample)
    if sample.condition_12:
        list2.append(sample)

# AND
for sample in samples:
    if sample.condition_1:
        list1.append(sample)

for sample in samples:
    if sample.condition_12:
        list2.append(sample)
```